exec/java-exec/src/main/java/org/apache/drill/exec/coord/zk/ZKACLProviderFactory.java

```diff
@@ -0,0 +1,113 @@
+            ZKACLProviderDelegate aclProvider = findACLProvider(config, stateProvider, context);
+            return aclProvider;
+        }
+        logger.trace("Using un-secure default ZK ACL");
+        final ZKDefaultACLProvider defaultAclProvider = new ZKDefaultACLProvider(stateProvider);
+        return new ZKACLProviderDelegate(defaultAclProvider);
+    }
+
+    public static ZKACLProviderDelegate findACLProvider(DrillConfig config, ZKACLContextProvider contextProvider,
+                                                        BootStrapContext context) throws DrillbitStartupException {
+        if (!config.hasPath(ZK_ACL_PROVIDER)) {
+            throw new DrillbitStartupException(String.format("BOOT option '%s' is missing in config.", ZK_ACL_PROVIDER));
+        }
+
+        final String aclProviderName = config.getString(ZK_ACL_PROVIDER);
+
+        if (Strings.isNullOrEmpty(aclProviderName)) {
+            throw new DrillbitStartupException(String.format("Invalid value '%s' for BOOT option '%s'", aclProviderName,
+                    ZK_ACL_PROVIDER));
+        }
+
+        ScanResult scan = context.getClasspathScan();
+        final Collection<Class<? extends ZKACLProvider>> aclProviderImpls =
+                scan.getImplementations(ZKACLProvider.class);
+        logger.debug("Found ZkACLProvider implementations: {}", aclProviderImpls);
+
+        for (Class<? extends ZKACLProvider> clazz : aclProviderImpls) {
+          final ZKACLProviderTemplate template = clazz.getAnnotation(ZKACLProviderTemplate.class);
+          if (template == null) {
+            logger.warn("{} doesn't have {} annotation. Skipping.", clazz.getCanonicalName(),
+                    ZKACLProviderTemplate.class);
+            continue;
+          }
+
+          if (template.type().equalsIgnoreCase(aclProviderName)) {
+            Constructor<?> validConstructor = null;
+            Class constructorArgumentClass = ZKACLContextProvider.class;
+            for (Constructor<?> c : clazz.getConstructors()) {
+              Class<?>[] params = c.getParameterTypes();
+              if (params.length == 1 && params[0] == constructorArgumentClass) {
+                validConstructor = c;
+                break;
+              }
+            }
+            if (validConstructor == null) {
+              logger.warn("Skipping ZKACLProvider implementation class '{}' since it doesn't " +
+                       "implement a constructor [{}({})]", clazz.getCanonicalName(), clazz.getName(),
+                      constructorArgumentClass.getName());
+              continue;
+            }
+            try {
+              final ZKACLProvider aclProvider = (ZKACLProvider) validConstructor.newInstance(contextProvider);
+              return new ZKACLProviderDelegate(aclProvider);
+            } catch (InvocationTargetException | IllegalArgumentException | IllegalAccessException | InstantiationException e ) {
```

Comment:
Is possible to replace all with `ReflectiveOperationException`.

Reply:
Thanks for pointing this out ! Done.